src/actors/symbolication.rs

```diff
@@ -640,54 +714,73 @@ fn symbolicate_frame(
     frame: &mut RawFrame,
     index: usize,
 ) -> Result<Vec<SymbolicatedFrame>, FrameStatus> {
-    let (object_info, symcache) = match caches.lookup_symcache(frame.instruction_addr.0) {
-        Some((_, info, Some(symcache))) => {
-            frame.package = info.raw.code_file.clone();
-            (info, symcache)
-        }
-        Some((_, info, None)) => {
-            frame.package = info.raw.code_file.clone();
-            if info.debug_status == ObjectFileStatus::Malformed {
+    let lookup_result = if let Some(lookup_result) =
+        caches.lookup_symcache(frame.instruction_addr.0, frame.addr_mode)
+    {
+        frame.package = lookup_result.object_info.raw.code_file.clone();
+        if lookup_result.symcache.is_none() {
+            if lookup_result.object_info.debug_status == ObjectFileStatus::Malformed {
                 return Err(FrameStatus::Malformed);
             } else {
                 return Err(FrameStatus::Missing);
             }
         }
-        None => return Err(FrameStatus::UnknownImage),
+        lookup_result
+    } else {
+        return Err(FrameStatus::UnknownImage);
     };
 
     log::trace!("Loading symcache");
-    let symcache = match symcache.parse() {
+    let symcache = match lookup_result
+        .symcache
+        .as_ref()
+        .expect("symcache should always be available at this point")
+        .parse()
+    {
         Ok(Some(x)) => x,
         Ok(None) => return Err(FrameStatus::Missing),
         Err(_) => return Err(FrameStatus::Malformed),
     };
 
-    let is_crashing_frame = index == 0;
-    let ip_register_value = if is_crashing_frame {
-        symcache
-            .arch()
-            .cpu_family()
-            .ip_register_name()
-            .and_then(|ip_reg_name| registers.get(ip_reg_name))
-            .map(|x| x.0)
-    } else {
-        None
-    };
-
-    let caller_address = InstructionInfo::new(symcache.arch(), frame.instruction_addr.0)
-        .is_crashing_frame(is_crashing_frame)
-        .signal(signal.map(|signal| signal.0))
-        .ip_register_value(ip_register_value)
-        .caller_address();
-
-    let relative_addr = match caller_address.checked_sub(object_info.raw.image_addr.0) {
-        Some(x) => x,
-        None => {
-            log::warn!("Underflow when trying to subtract image start addr from caller address");
-            metric!(counter("relative_addr.underflow") += 1);
-            return Err(FrameStatus::MissingSymbol);
+    // get the relative caller address
+    let relative_addr = if let Some(addr) = lookup_result.relative_addr {
```

Comment:
_nit_: Since `SymcacheLookupResult` is now a thing, this could become a method on the result type.

Reply:
I tried this now but it takes too many args. Hmm